batcher.go

```diff
@@ -0,0 +1,304 @@
+func (b *Batcher[K, V]) batchTicker() {
+	limiter := rate.NewLimiter(rate.Every(b.opt.Ticker), 1)
+	for limiter.Wait(b.ctx) == nil {
+		func() {
+			b.mu.Lock()
+			defer b.mu.Unlock()
+
+			b.runBatch(TimerBatcher)
+		}()
+	}
+}
+
+type waiter[K comparable, V any] struct {
+	ctx context.Context
+	key K
+	res chan *Result[V]
+}
+
+// Load load data with key and return value and error
+func (b *Batcher[K, V]) Load(ctx context.Context, key K) (V, error) {
+	if b.ctx == nil {
+		var v V
+
+		return v, ErrBatcherNotInit
+	}
+
+	ctx, cancel := b.context(ctx)
+	defer cancel()
+
+	waiter := b.load(ctx, key)
+
+	select {
+	case res := <-waiter:
+		return res.Value, res.Err
+	case <-ctx.Done():
+		var v V
+		return v, ctx.Err()
+	}
+}
+
+// LoadMany load slice data and return value and error map
+func (b *Batcher[K, V]) LoadMany(ctx context.Context, keys ...K) (map[K]V, map[K]error) {
+	data := make(map[K]V, len(keys))
+	errors := make(map[K]error, len(keys))
+
+	if b.ctx == nil {
+		for _, key := range keys {
+			errors[key] = ErrBatcherNotInit
+		}
+
+		return data, errors
+	}
+
+	ctx, cancel := b.context(ctx)
+	defer cancel()
+
+	waiterList := make(map[K]<-chan *Result[V], len(keys))
+	var ok bool
+	for _, key := range keys {
+		if _, ok = waiterList[key]; ok {
+			continue
+		}
+
+		waiterList[key] = b.load(ctx, key)
+	}
+
+	// Need read all channel
+	for key, wt := range waiterList {
+		select {
+		case res := <-wt:
+			if res.Err != nil {
+				errors[key] = res.Err
+			} else {
+				data[key] = res.Value
+			}
+		case <-ctx.Done():
+			errors[key] = ctx.Err()
+		}
+	}
+
+	return data, errors
+}
+
+func (b *Batcher[K, V]) load(ctx context.Context, key K) <-chan *Result[V] {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	c := make(chan *Result[V], 1)
+
+	if b.shutdown {
+		c <- &Result[V]{Err: errors.New("batcher: shut down")}
+		close(c)
+
+		return c
+	}
+
+	b.pending = append(b.pending, waiter[K, V]{ctx: ctx, key: key, res: c})
+
+	b.runBatch(MainBatcher)
+
+	return c
+}
+
+// runBatch checks the number of valid threads and pending data and starts a new thread to process the data.
+// b.mu must be lock before
+func (b *Batcher[K, V]) runBatch(btype batcherType) {
+	// If the number of threads is the maximum, we do not create a new one.
+	// We do not create a batcher "TimerBatcher", because the maximum number of threads shows
+	// the optimal operation of the system (all keys are taken)
+	if b.nHandlers >= b.opt.MaxHandlers {
+		return
+	}
+
+	batch := b.nextBatch(btype)
+	if len(batch) > 0 {
+		b.wg.Add(1)
+		go func() {
+			b.callHandlers(btype, batch)
+			b.wg.Done()
+		}()
+		b.nHandlers++
+	}
+}
+
+// nextBatch fetches data from pending.
+// b.mu must be lock before
+func (b *Batcher[K, V]) nextBatch(btype batcherType) (bch []waiter[K, V]) {
+	if len(b.pending) < b.opt.MinBatchSize && btype != TimerBatcher {
+		return nil
+	}
+
+	// Send all
+	if b.opt.MaxBatchSize == 0 || len(b.pending) <= b.opt.MaxBatchSize {
```

Comment:
Maybe I just don't get it, but looks like you only set `b.pending` on lines 211 and 225, below this line. So, this if clause will always result in `true` and the latter code will never be executed. 🤔